lib/rpc/handler/FallbackHandler.ts

```diff
@@ -0,0 +1,97 @@
+import Logger from 'bunyan'
+import { ChainId } from '@uniswap/sdk-core'
+import { getProviderId } from '../utils'
+import { ProviderHealthStateRepository } from '../ProviderHealthStateRepository'
+import { ProviderHealthStateDynamoDbRepository } from '../ProviderHealthStateDynamoDbRepository'
+import { ProviderHealthState } from '../ProviderHealthState'
+import { MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { metricScope, MetricsLogger } from 'aws-embedded-metrics'
+import { APIGatewayProxyResult } from 'aws-lambda'
+import { AWSMetricsLogger } from '../../handlers/router-entities/aws-metrics-logger'
+
+interface AlarmEvent {
+  alarmName: string
+  state: string
+  previousState: string
+  providerId: string
+  reason: string
+}
+
+export class FallbackHandler {
+  private healthStateRepository: ProviderHealthStateRepository
+  private log: Logger
+
+  constructor(log: Logger) {
+    const dbTableName = process.env.PROVIDER_HEALTH_STATE_DB_TABLE_NAME
+    if (!dbTableName) {
+      throw new Error('Missing DB_TABLE_NAME in env var')
+    }
+    this.healthStateRepository = new ProviderHealthStateDynamoDbRepository(dbTableName, log)
+    this.log = log
+  }
+
+  private buildHandler() {
+    return metricScope((metricsLogger: MetricsLogger) => async (event: object): Promise<APIGatewayProxyResult> => {
+      metricsLogger.setNamespace('Uniswap')
+      metricsLogger.setDimensions({ Service: 'RoutingAPI' })
+      const metric = new AWSMetricsLogger(metricsLogger)
+
+      const alarmEvent = this.readAlarmEvent(event)
+      this.log.debug({ alarmEvent }, 'Parsed alarmEvent')
+
+      if (
+        (alarmEvent.previousState === 'OK' || alarmEvent.previousState === 'INSUFFICIENT_DATA') &&
+        alarmEvent.state === 'ALARM'
+      ) {
+        metric.putMetric(`RPC_GATEWAY_FALLBACK_${alarmEvent.providerId}_INTO_UNHEALTHY`, 1, MetricLoggerUnit.Count)
+        this.log.error(
+          `${alarmEvent.providerId} becomes UNHEALTHY due to ${alarmEvent.previousState}=>ALARM in ${alarmEvent.alarmName}`
+        )
+        await this.healthStateRepository.write(alarmEvent.providerId, ProviderHealthState.UNHEALTHY)
+      } else if (
+        alarmEvent.previousState === 'ALARM' &&
+        (alarmEvent.state === 'OK' || alarmEvent.state === 'INSUFFICIENT_DATA')
```

Comment:
same comment about the previous state. In this case, if the previous state was insufficient_data, then we would not turn back to green. (even though it the transition from ALARM -> INSUFFICIENT_DATA is captured by this `if` the transition from OK to INSUFFICIENT_DATA is not)

Reply:
It's a good question to discuss. There's no doubt about how we treat `OK` or `ALARM`, but for `INSUFFICIENT_DATA` it's up to some debate.

If we are being optimistic, we should treat `INSUFFICIENT_DATA` as benign. I think this aligns more accurately with our expected behaviors: When the services starts, it has no knowledge about provider's health state, but it will treat the provider as benign and use it anyway.

@cgkol what do you think about it? Can you share your comments?

Reply:
I would rather that we change on insufficient_data, becuase I'd rather we dont get stuck on a bad state. However, given that when it starts it might take a while for it to get healthy, I think thats fair to just not change anything with `INSUFFICIENT_DATA`, just make sure its well document on the comments; assuming insufficient_data m

Reply:
i think wheneve you ahve insufficient_data, default to previous state

Reply:
Thanks for the suggestion. The logic in `FallbackHandler.ts` are now fixed.